messaging/app/src/main/java/com/google/firebase/example/messaging/kotlin/MainActivity.kt

```diff
@@ -130,4 +139,41 @@ class MainActivity : AppCompatActivity() {
     }
     // [END ask_post_notifications]
 
+    // [START get_store_token]
+    private suspend fun getAndStoreRegToken(): String {
+        val token = Firebase.messaging.token.await()
+        // Add token and timestamp to Firestore for this user
+        val deviceToken = hashMapOf(
+            "token" to token,
+            "timestamp" to FieldValue.serverTimestamp(),
+        )
+
+        // Get user ID from Firebase Auth or your own server
+        Firebase.firestore.collection("fcmTokens").document("myuserid")
+            .set(deviceToken).await()
+        return token
+    }
+    // [END get_store_token]
+
+    private fun refreshToken() {
+        // In the app’s first Activity
+        val preferences = this.getPreferences(Context.MODE_PRIVATE)
+        val lastRefreshLong = preferences.getLong("lastRefreshDate", -1)
+        lifecycleScope.launch {
+            val today = Date()
+            val c = Calendar.getInstance().apply {
+                time = if (lastRefreshLong == -1L) today else Date(lastRefreshLong)
+                add(Calendar.DATE, 30)
+            }
+
+            if (today.after(c.time) || lastRefreshLong == -1L) {
+                // get token and store into Firestore
+                getAndStoreRegToken()
+                // sync device cache time with Firestore just in case
+                val document = Firebase.firestore.collection("refresh").document("refreshDate").get().await()
+                val updatedTime = (document.data!!["lastRefreshDate"] as Timestamp).seconds * 1000
+                preferences.edit().putLong("lastRefreshDate", updatedTime).apply()
+            }
+        }
+    }
 }
```

Comment:
@andreaowu This function had no snippets, but I wonder if we want/should show this logic on the doc somewhere? Otherwise I'll remove it from this PR and we can keep it on the quickstart only.

Reply:
Yeah, I've thought about this too. This is an optimization basically that saves the app from needing to call Firestore every time the app opens - it'll first check the cache to see if today's date is 30 days beyond the cached date, and if so, it'll grab the date from Firestore and update the cached date. I don't think this needs to go in the snippets, I've explained it in the blogpost